### PR TITLE
feat(packageRules): migrate matchers and excludes

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Migrate matchPackageX and excludePackageX plus associated matchDepX to use matchPackageNames and matchDepNames.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Blocked by #28591

File: lib/util/package-rules/index.ts [hunk 1]
```diff
@@ -4,22 +4,16 @@ import { mergeChildConfig } from '../../config';
...
 import { logger } from '../../logger';
 import matchers from './matchers';
-import { matcherOR } from './utils';
 
 function matchesRule(
   inputConfig: PackageRuleInputConfig,
   packageRule: PackageRule,
 ): boolean {
-  let positiveMatch = true;
+  const positiveMatch = true;
```
Comment (secustor) [inline lib/util/package-rules/index.ts:12]: If this is now a const than it is no longer needed

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/util/package-rules/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/package-rules/index.ts:12]: Ha, VSCode/eslint autochanged that, I didn't notice

File: lib/modules/datasource/npm/npmrc.ts [hunk 2]
```diff
@@ -168,10 +168,10 @@ export function setNpmrc(input?: string): void {
 export function resolveRegistryUrl(packageName: string): string {
   let registryUrl = defaultRegistryUrls[0];
   for (const rule of packageRules) {
-    const { matchPackagePrefixes, registryUrls } = rule;
+    const { matchPackageNames, registryUrls } = rule;
     if (
-      !matchPackagePrefixes ||
-      packageName.startsWith(matchPackagePrefixes[0])
+      !matchPackageNames ||
+      packageName.startsWith(matchPackageNames[0].replace(/\*\*$/, ''))
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/npm/npmrc.ts:174]: We use regEx to protect ourselves from bad regular expressions. Shouldn't be needed here for such a simple match?

[the remark this answers is not in the document]
Comment (secustor) [reply lib/modules/datasource/npm/npmrc.ts:174]: We use it tough also to cache it IIRC.

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -3044,7 +2901,7 @@ For more details on supported syntax see Renovate's [string pattern matching doc
       "groupName": "high merge confidence"
     }
   ]
-}
+##}
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2904]:
```suggestion
}
```

File: lib/config/validation.ts [hunk 4]
```diff
@@ -562,18 +550,13 @@ export async function validateConfig(
...
-                'matchDepPatterns',
-                'excludeDepPatterns',
-              ].includes(key)
-            ) {
+            if (['matchPackageNames', 'matchDepNames'].includes(key)) {
+              const startPattern = regEx(/!?\//);
+              const endPattern = regEx(/\/g?i?$/);
               for (const pattern of val as string[]) {
-                if (pattern !== '*') {
+                if (pattern.match(startPattern) && pattern.match(endPattern)) {
```
Comment (viceice) [inline lib/config/validation.ts:557]:
```suggestion
                if (startPattern.test(pattern) && endPattern.test(pattern)) {
```

File: lib/config/validation.ts [hunk 5]
```diff
@@ -562,18 +550,13 @@ export async function validateConfig(
...
                   try {
-                    regEx(pattern);
+                    regEx(pattern.replace(startPattern, '/'));
```
Comment (viceice) [inline lib/config/validation.ts:559]: why the replace? don't we need to remove the end pattern too?

File: lib/util/package-rules/repositories.spec.ts [hunk 6]
```diff
@@ -1,11 +1,11 @@
 import { RepositoriesMatcher } from './repositories';
 
 describe('util/package-rules/repositories', () => {
-  const packageNameMatcher = new RepositoriesMatcher();
+  const repositoryMatcher = new RepositoriesMatcher();
```
Comment (viceice) [inline lib/util/package-rules/repositories.spec.ts:4]: this rename makes a lot noise, we should do it in current main before

File: docs/usage/faq.md [hunk 7]
```diff
@@ -97,7 +97,7 @@ The basic idea is that you create a new `packageRules` entry and describe what k
 {
   "packageRules": [
     {
-      "matchPackagePatterns": ["^jest"],
+      "matchPackageNames": ["jest"],
```
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/faq.md:100]: Some of the examples seemed poorly chosen so I didn't always map them exactly

File: lib/util/package-rules/repositories.spec.ts (hunk 6 above)
Comment (rarkins) [reply lib/util/package-rules/repositories.spec.ts:4]: https://github.com/renovatebot/renovate/pull/28652

File: lib/config/validation.ts (hunk 5 above)
Comment (rarkins) [reply lib/config/validation.ts:559]: `regEx() doesn't recognize our `!/..../` syntax but should be fine with the end syntax. Hence I replace any leading `!/` with `/` but leave the trailing as-is

**Review by viceice — COMMENTED** [review]
needs rebase

**Review by viceice — DISMISSED** [review]
we need to check again when we change presets in main (like I did recently)

File: docs/usage/configuration-options.md [hunk 8]
```diff
@@ -2939,9 +2841,8 @@ The above will configure `rangeStrategy` to `pin` only for the npm package `angu
 {
   "packageRules": [
     {
-      "matchDatasources": ["npm"],
-      "matchPackageNames": ["@angular/*", "!@angular/abc"],
-      "groupName": "Angular"
+      "matchPackagePatterns": ["^angular"],
+      "rangeStrategy": "replace"
```
Comment (secustor) [inline docs/usage/configuration-options.md:2845]:
```suggestion
      "matchDatasources": ["npm"],
      "matchPackageNames": ["@angular/*", "!@angular/abc"],
      "groupName": "Angular"
```

> The above will group together any npm package which starts with `@angular/` except `@angular/abc`.

https://github.com/renovatebot/renovate/pull/28602/files#diff-be0a39f342380493f17c380bbd694f1f05d18bdd8d7cecf174369084952102deR2851
This does not fit the description

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -3142,7 +2998,7 @@ For example, the following package rule can be used to replace the registry for
   "packageRules": [
     {
       "matchDatasources": ["docker"],
-      "matchPackagePatterns": ["^docker\\.io/.+"],
+      "matchPackageNames": ["docker\\.io/.+**"],
```
Comment (secustor) [inline docs/usage/configuration-options.md:3001]:
```suggestion
      "matchPackageNames": ["docker.io/**"],
```
this is a mix of regex and glob

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -3157,13 +3013,13 @@ Or, to add a registry prefix to any `docker` images that do not contain an expli
     {
       "description": "official images",
       "matchDatasources": ["docker"],
-      "matchPackagePatterns": ["^[a-z-]+$"],
+      "matchPackageNames": ["[a-z-]+$**"],
```
Comment (secustor) [inline docs/usage/configuration-options.md:3016]:
```suggestion
      "matchPackageNames": ["/^[a-z-]+$/"],
```

File: docs/usage/configuration-options.md [hunk 11]
```diff
@@ -3157,13 +3013,13 @@ Or, to add a registry prefix to any `docker` images that do not contain an expli
...
       "replacementNameTemplate": "some.registry.org/library/{{{packageName}}}"
     },
     {
       "description": "non-official images",
       "matchDatasources": ["docker"],
-      "matchPackagePatterns": ["^[a-z-]+/[a-z-]+$"],
+      "matchPackageNames": ["[a-z-]+/[a-z-]+$**"],
```
Comment (secustor) [inline docs/usage/configuration-options.md:3022]:
```suggestion
      "matchPackageNames": ["/^[a-z-]+/[a-z-]+$/"],
```

**Review by secustor — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 8 above)
Comment (rarkins) [reply docs/usage/configuration-options.md:2845]: I thought it does, how come you think why not?

Comment (secustor) [reply docs/usage/configuration-options.md:2845]: The description is below not above the code snippet. Notice the `The above will` prefixes.

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**